lib/RestructureCFG/SelectScopePass.cpp

```diff
@@ -0,0 +1,261 @@
+              "The identified postdominator is "
+                << PostDominator->getName().str() << "\n");
+
+    SmallVector<BasicBlock *>
+      NodesToProcess = getNodesToProcess(ConditionalNode, PostDominator);
+    revng_log(SelectScopePassLogger,
+              "Nodes between conditional and its postdominator, in reverse "
+              "post order:\n");
+    for (auto DFSNode : NodesToProcess) {
+      revng_log(SelectScopePassLogger, "  " << DFSNode->getName().str());
+    }
+
+    // Initialize the `ReachabilityMap`
+    std::map<BasicBlock *, size_t> ReachabilityMap;
+
+    // We initialize the `ReachabilityMap` for each `ConditionalSuccessor`
+    for (const auto &[Index, ConditionalSuccessor] :
+         enumerate(ConditionalSuccessors)) {
+      if (not ReachabilityMap.contains(ConditionalSuccessor)) {
+        ReachabilityMap[ConditionalSuccessor] = Index;
+      }
+    }
+
+    // Process each node in the zone of interest
+    for (BasicBlock *Candidate : NodesToProcess) {
+      revng_log(SelectScopePassLogger,
+                "Analyzing candidate: " + Candidate->getName().str() << "\n");
+
+      // We precompute the predecessors to avoid invalidation due to graph
+      // changes. It is fundamental that we always traverse the `ScopeGraph`
+      // view of the CFG, or we may end up with some inconsistencies in terms
+      // of the visited nodes.
+      revng_log(SelectScopePassLogger, "The candidate predecessors are:\n");
+      SetVector<BasicBlock *> Predecessors = getPredecessors(Candidate);
+
+      // `Candidate`, could be itself a immediate successor of a conditional
+      // node, and therefore correspond to a `ScopeID`. We therefore need to
+      // take into consideration it when assigning the final scope for each
+      // `Candidate`.
+      // We can do this by always enqueuing `Candidate` as a predecessor of
+      // itself, this can lead to two situations:
+      // 1) `Candidate` is not a successor of the conditional, therefore no
+      //    corresponding entry in `ReachabilityMap` will be present, and this
+      //    will not influence the decision on the `ScopeID` which will be
+      //    finally assigned.
+      // 2) `Candidate` is a successor of the conditional, therefore a
+      //    corresponding entry in `ReachabilityMap` will be present, and it
+      //    will be correctly taken into account for the `ScopeID` decision
+      //    process.
+      // Alternatively, we could pre-assign the `ScopeID`, in the
+      // `ReachabilityMap`, for each successor of a conditional node during
+      // the initialization. This, however, would tie us to the decision of
+      // always assigning the successor of a conditional node to the `ScopeID`
+      // opening in the successor itself, while, in principle, we could
+      // alternatively disconnect the edge connecting the conditional and the
+      // successor, by making it a `goto` edge.
+      // Predecessors.push_back(Candidate);
```

Comment:
The last line here is a leftover.

Reply:
Nice catch, this has been dropped here https://github.com/revng/revng/blob/a3b1cbac550057e016f37c8f9eb201e8ab290b82/lib/RestructureCFG/SelectScopePass.cpp#L168